[stdlib] Add `Span.find()` and `rfind()`

martinvuyk opened the pull request:
Add `Span.find()` and `rfind()`. This will be useful to look for raw data, not only string processing.

JoeLoser commented on stdlib/src/utils/span.mojo:
```diff
@@ -335,3 +338,194 @@ struct Span[
             unsafe_ptr=self._data, len=self._len
         )
+
+    fn find[
```
**Question** It seems a bit weird from an API design perspective to have this "find" function in `span`, can you help me understand why we'd want this?  It feels more coupled to string algorithms, and `Span` isn't just for a view over string data.

martinvuyk replied:
It can be used if you want to find occurrences of a specific scalar value sequence. My main thought was that we'll lose the ability to do that once `StringSlice.find()` is fixed to use unicode codepoints and it introduces quite some overhead for algorithms that are faster using raw bytes. We could also make this private WDYT?

martinvuyk replied:
@JoeLoser gentle ping and also to add something, since it seems like we are moving toward unifying Python's `bytes` into `Span[UInt8]`in PR #3636, request #3634 and many other places. Python's [bytes.find](https://docs.python.org/3/library/stdtypes.html#bytes.find) is actually a method, and many other things that we put in `StringSlice` and `StringRef` are actually [operations from `bytes`](https://docs.python.org/3/library/stdtypes.html#bytes-and-bytearray-operations).

msaelices replied:
I think this is a great change for overall performance when manipulating slices. It opens a lot of possibilities.

owenhilyard replied:
@JoeLoser per discussion on discord, `Span[T].find` is intended to be `memchr` or `memmem` as appropriate, but that the subspan of bytes must be aligned to `alignof[T]()`. We can probably expand it to any `T: Trivial` once we have a `Trivial` trait. 

The proper string version likely needs be specialized for UTF-8 comparisons. @martinvuyk and I have discussed an "AsciiString" which can more aggressively optimize for using byte-oriented intrinsics since UTF-8 is annoying to do with SIMD.

martinvuyk commented:
@JoeLoser CI is failing because of unrelated reasons and ignore the diff numbers, it's because of the magic.lock file. 

I refactored `normalize_index` and used it here. I also refactored `memrmem` and `memrchr` to be vectorized like their counterparts. These are algorithms that apply not only to string data, you can use them to find anything (DB indexes etc.). I had tried to make a similar algorithm for Array in my repo but failed to make it as efficient as these, kudos to whoever implemented them. Can we please have this? They will be used a lot. I've also found that `.replace()` and `.count()` will use this function besides `.split()`, `.splitlines()`, `List.index()`, and others.

martinvuyk commented:
@JoeLoser would it be possible to land this before the next stable release? I'd like to use this to be able to bypass string's `.find()` for algorithms which work on raw bytes. And using that bypass, I'd like to make the breaking change to switch over to full unicode:
- `String.__len__()` should return unicode length and not byte length
- `String.find()` should work on unicode offsets
- `String.__getitem__()` should work on unicode offsets

I would like to make the change as soon as possible because it will break a lot of projects, and many more are coming because the community is growing.

martinvuyk commented:
@JoeLoser we need this to be able to support unicode indexing while also keeping the ability to look for raw bytes.

gryznar commented:
Shall we follow Python convention? `rfind` is not obvious at first glance without diving into documentation

martinvuyk commented:
> Shall we follow Python convention? rfind is not obvious at first glance without diving into documentation

I have no problem going for `find_from_right` or something like that. But let's leave it for another PR

gryznar commented:
This question is to @JoeLoser if stdlib team is open to cleaning this up in the nearest future, step by step

JoeLoser commented on mojo/stdlib/stdlib/memory/span.mojo:
```diff
@@ -626,3 +632,165 @@ struct Span[
+
+        @parameter
+        if from_left and not single_value:
+            loc = _memmem(
```
**Comment** Can't wait to land your other PRs that change these APIs to use a safer `Span` soon.

JoeLoser commented on mojo/stdlib/test/collections/string/test_string.mojo:
```diff
@@ -631,6 +635,24 @@ def test_find():
+        assert_equal(51, (String.ASCII_LETTERS * i).find("Za"))
+
+    # Special characters.
+    # FIXME: once find works by unicode codepoints
```
**Suggestion** Link to the appropriate GitHub issue, please.

JoeLoser reviewed (approved):
Generally looks good, just needs a rebase.  We're +1 for adding these algorithms to `Span` as discussed in https://github.com/modular/modular/issues/4758 - we discussed this issue as a team the other day.

Thanks, Martin!

martinvuyk commented on mojo/stdlib/stdlib/memory/span.mojo:
```diff
@@ -626,3 +632,165 @@ struct Span[
+
+        # _memXXX implementations already handle when haystack_len == 0
+        @parameter
+        if unsafe_dont_normalize:
```
The logic is very different, python's find clamps indices to accept any range. I opened  #3783 a very long time ago because there are several functions like this with different clamping or normalization criteria, some also accept or reject containers with 0 length etc.

gryznar replied:
Should we strictly follow Python here? With extensions, I suppose that implementation in Mojo could be different

JoeLoser commented:
!sync

martinvuyk committed:
Add Span.find() and rfind()

Signed-off-by: martinvuyk <martin.vuyklop@gmail.com>

martinvuyk committed:
Add Span.find() and rfind()

Signed-off-by: martinvuyk <martin.vuyklop@gmail.com>
